compiler/rustc_ast_lowering/src/local.rs

```diff
@@ -0,0 +1,68 @@
+use super::LoweringContext;
+
+use rustc_ast::*;
+use rustc_data_structures::fx::FxIndexMap;
+use rustc_data_structures::thin_vec::ThinVec;
+use rustc_hir as hir;
+use rustc_session::parse::feature_err;
+use rustc_span::symbol::Ident;
+use rustc_span::{sym, DUMMY_SP};
+
+impl<'hir> LoweringContext<'_, 'hir> {
+    /// Before: `let Some([ref x, ref y]) = z else { return }`
+    /// After: `let (ref x, ref y) = match z { Some([x, y]) => (x, y), _ => return }`
+    pub(super) fn lower_local_let_else(
+        &mut self,
+        l: &Local,
+        scrutinee: &Expr,
+        els: &Block,
+    ) -> (&'hir hir::Pat<'hir>, &'hir hir::Expr<'hir>) {
+        let (then_pat, bindings) = {
+            self.let_else_bindings = Some(FxIndexMap::default());
```

Comment:
You need to save `let_else_bindinds` and to restore it later.

Reply:
Hmm I thought I was guaranteed not to have any locals within patterns, and so this would not be hit recursively? If that is not the case, maybe `let_else_bindings` needs to be a stack instead of an `Option`.

Reply:
Actually I can just do what you suggested and be safe in any case.